Highlight spans with errors

virtuald opened the pull request:
- Annotations: value must be 'error'
- Binary annotations: key must be 'error'
- Fixes #1140 

Here's three screenshots that illustrate the changes:

Traces view (top is transient, bottom is critical):
![traces](https://cloud.githubusercontent.com/assets/567900/16543254/88b323de-4099-11e6-8b7c-eca266a5ce74.png)

Trace view (spans are critical, normal, critical + error annotation, transient):
![trace](https://cloud.githubusercontent.com/assets/567900/16543257/8fdc06f8-4099-11e6-9ba5-580c0eafc0d3.png)

Trace modal:
![trace-modal](https://cloud.githubusercontent.com/assets/567900/16543258/976f442a-4099-11e6-8d6f-4dbd0131c622.png)

And the JSON to duplicate:

```
[
    {
        "annotations": [
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server"
                },
                "timestamp": 1466977002899000,
                "value": "sr"
            },
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server"
                },
                "timestamp": 1466977002908000,
                "value": "ss"
            }
        ],
        "binaryAnnotations": [
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server"
                },
                "key": "http.status_code",
                "value": "500"
            },
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server"
                },
                "key": "http.url",
                "value": "/error"
            },
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server"
                },
                "key": "error",
                "value": "You done messed up son"
            }
        ],
        "duration": 9000,
        "id": "8e5b31230ccb7f42",
        "name": "get",
        "timestamp": 1466977002899000,
        "traceId": "8e5b31230ccb7f42"
    },
    {
        "annotations": [
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server"
                },
                "timestamp": 1466977002899000,
                "value": "sr"
            },
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server"
                },
                "timestamp": 1466977002908000,
                "value": "ss"
            }
        ],
        "binaryAnnotations": [
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server"
                },
                "key": "http.status_code",
                "value": "500"
            },
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server"
                },
                "key": "http.url",
                "value": "/error"
            }
        ],
        "duration": 9000,
        "id": "8e5b31230ccb7e30",
        "parentId": "8e5b31230ccb7f42",
        "name": "getit",
        "timestamp": 1466977002899000,
        "traceId": "8e5b31230ccb7f42"
    },
    {
        "annotations": [
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server"
                },
                "timestamp": 1466977002899000,
                "value": "sr"
            },
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server"
                },
                "timestamp": 1466977002908000,
                "value": "ss"
            },
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server"
                },
                "timestamp": 1466977002907000,
                "value": "error"
            }
        ],
        "binaryAnnotations": [
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server"
                },
                "key": "http.status_code",
                "value": "500"
            },
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server"
                },
                "key": "http.url",
                "value": "/error"
            },
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server"
                },
                "key": "error",
                "value": "You done messed up son"
            }
        ],
        "duration": 9000,
        "id": "8e5b31230ccb7e43",
        "parentId": "8e5b31230ccb7e30",
        "name": "gettt",
        "timestamp": 1466977002899000,
        "traceId": "8e5b31230ccb7f42"
    },
    {
        "annotations": [
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server2"
                },
                "timestamp": 1466977002899000,
                "value": "sr"
            },
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server2"
                },
                "timestamp": 1466977002908000,
                "value": "ss"
            },{
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server2"
                },
                "timestamp": 1466977002907000,
                "value": "error"
            }
        ],
        "binaryAnnotations": [
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server2"
                },
                "key": "http.status_code",
                "value": "500"
            },
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server2"
                },
                "key": "http.url",
                "value": "/error"
            }
        ],
        "duration": 19000,
        "id": "8e5b31230ccb7e53",
        "parentId": "8e5b31230ccb7e43",
        "name": "geterdone",
        "timestamp": 1466977002899000,
        "traceId": "8e5b31230ccb7f42"
    },
    {
        "annotations": [
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server2"
                },
                "timestamp": 1466977002899000,
                "value": "sr"
            },
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server2"
                },
                "timestamp": 1466977002908000,
                "value": "ss"
            },{
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server2"
                },
                "timestamp": 1466977002907000,
                "value": "error"
            }
        ],
        "binaryAnnotations": [
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server2"
                },
                "key": "http.status_code",
                "value": "500"
            },
            {
                "endpoint": {
                    "ipv4": "10.24.23.10",
                    "port": 9411,
                    "serviceName": "zipkin-server2"
                },
                "key": "http.url",
                "value": "/error"
            }
        ],
        "duration": 19000,
        "id": "8e5b31230ccb7e99",
        "name": "is-transient",
        "timestamp": 1466977002899000,
        "traceId": "8e5b31230ccb7e99"
    }
]
```

codefromthecrypt commented on zipkin-ui/js/component_ui/traceSummary.js:
```diff
@@ -124,6 +124,13 @@ function getSpanTimestamps(spans) {
   }))).value();
 }
 
+function detectErrors(spans) {
```
possible add a unit test for this?

codefromthecrypt commented on zipkin-ui/js/component_ui/traceToMustache.js:
```diff
@@ -106,8 +106,12 @@ export default function traceToMustache(trace) {
       const spanStartTs = span.timestamp || traceTimestamp;
       const spanDepth = spanDepths[span.id] || 1;
       const width = (span.duration || 0) / summary.duration * 100;
+      let isError = false;
```
not sure, but I _think_ this function is unit tested, if accessible, mind adding a test for 'error' -> isError?

virtuald replied:
I'll look into it tomorrow.

eirslett replied:
You can do const isError = a.key === 'error' to avoid the let.

virtuald replied:
@eirslett but a.key is inside of a closure, I don't think you will be able to reassign the const each time.

eirslett replied:
Oh, I missed that, sorry!

codefromthecrypt commented:
if you can, please add ERROR to this and use the constant. we can change the java, thrift sides in a different PR.

codefromthecrypt commented:
this :) https://github.com/openzipkin/zipkin/blob/master/zipkin-ui/js/component_ui/traceConstants.js

codefromthecrypt commented:
something to consider.. if we think of an "error" event as/annotation transient, we might want to stain the span yellow instead. Presence of the binary annotation could imply complete failure (something the client would be responsible for in an RPC span, or the server is responsible for in a server-only root span).

See https://github.com/openzipkin/zipkin-api/pull/18/files/1dc521c1994c104211f9a72400e48eae5e948549#r68687632 from @yurishkuro

codefromthecrypt commented:
for example, I think finagle logs "finagle.retry" which is a stack-specific way of saying there was a transient error

virtuald commented:
I'll try to get to the adjustments tomorrow, no time tonight.

codefromthecrypt commented:
all your help has been appreciated, Dustin!

jcarres-mdsol commented:
There should be some documentation so client implementors know that this functionality is available

virtuald committed:
Highlight spans with errors

- Annotations: value must be 'error'
- Binary annotations: key must be 'error'
- Fixes #1140

virtuald commented:
Updated code to include concept of transient (has annotation) vs critical (has binary annotation) errors, added unit tests, updated screenshots, updated JSON.

codefromthecrypt commented:
great. will try it out soon!

codefromthecrypt commented:
tried and looks good. thanks!
